### DISCUSSION TITLE
Si4432 Init error

### DISCUSSION THREAD

**Discussion by radio-satellites** [jgromes/RadioLib#1277] [Q&A]
Hello!

I've built a custom PCB based on the ESP8266 and the Si4432. Here is my code for initialization, but the init fails. Is it a programming problem or a hardware one? I'll post the schematic in the next post. 

```c++
/*
  RadioLib Si443x Blocking Transmit Example

  This example transmits packets using Si4432 FSK radio module.
  Each packet contains up to 64 bytes of data, in the form of:
  - Arduino String
  - null-terminated char array (C-string)
  - arbitrary binary data (byte array)

  Other modules from Si443x/RFM2x family can also be used.

  Using blocking transmit is not recommended, as it will lead
  to inefficient use of processor time!
  Instead, interrupt transmit is recommended.

  For default module settings, see the wiki page
  https://github.com/jgromes/RadioLib/wiki/Default-configuration#si443xrfm2x

  For full API reference, see the GitHub Pages
  https://jgromes.github.io/RadioLib/
*/

// include the library
#include <RadioLib.h>
#include <SPI.h>

// Si4432 has the following connections:
// nSEL pin:  15
// nIRQ pin:  4
// SDN pin:   16
Si4432 radio = new Module(15, 4, 16);

// or using RadioShield
// https://github.com/jgromes/RadioShield
//Si4432 radio = RadioShield.ModuleA;

void setup() {
  Serial.begin(9600);

  // initialize Si4432 with default settings
  Serial.print(F("[Si4432] Initializing ... "));
  SPI.pins(14,12,13,15);
  SPI.begin();
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(10); }
  }
}

// counter to keep track of transmitted packets
int count = 0;

void loop() {
  Serial.print(F("[Si4432] Transmitting packet ... "));

  // you can transmit C-string or Arduino string up to
  // 64 characters long
  String str = "Hello World! #" + String(count++);
  int state = radio.transmit(str);

  // you can also transmit byte array up to 64 bytes long
  /*
    byte byteArr[] = {0x01, 0x23, 0x45, 0x56, 0x78, 0xAB, 0xCD, 0xEF};
    int state = radio.transmit(byteArr, 8);
  */

  if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F(" success!"));

  } else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
    // the supplied packet was longer than 256 bytes
    Serial.println(F(" too long!"));

  } else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
    // timeout occured while transmitting packet
    Serial.println(F(" timeout!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }

  // wait for a second before transmitting again
  delay(1000);
}

**Comment by radio-satellites** [comment]
Schematic:

![image](https://github.com/user-attachments/assets/4dba39c8-4aad-448e-b4c8-9325d0aa31e8)

**Comment by jgromes** [comment]
"Init fails" is not really a sufficient description - fails with what error code? Have tried to examine the debug output?

Also, since oyu have not passed your SPI isntance to the `Module` constructor, RadioLib will try to initialize it, which will lead to `SPI.begin` being called twice. I don't know if that will be a problem on your platform, but you should follow the wiki page that describes how to use non-standard SPI setup: https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup

**Reply by radio-satellites** [reply]
Hi Jan,

My apologies. I did write this up rather quickly. 

It looks like the SPI communication fails on .begin(). I'm getting the following output:

```
[Si4432] Initializing ... 
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
RLB_SPI: R	1	0	
failed, code -2
```

**Reply by jgromes** [reply]
The SPI seems to be always returning 0, so either your SPI is not configured/connected correctly, or the radio is not powered.

**Reply by radio-satellites** [reply]
Interesting, thanks for your reply! 

I tried tapping the GPIO12 pin randomly to VCC, hoping to check whether or not it was actually reading from the pin when it read the registers. I would assume something like that would have created random noise, but it was still all zeros. Does this indicate that it's simply not reading from the right SPI port?

**Reply by jgromes** [reply]
SPI read form one register takes a few dozen us at the default SPI speed - it is very unlikely that by randomly touching wires, you will hit it at the right moment. You would have to permanently connect it to the votlage (preferrably via a pullup, not a direct short). However, the preferrable way to check the hardware is correct is to check the MOSI/MISO/SCK and all other signals with an oscilloscope.

**Reply by radio-satellites** [reply]
Looks like it didn't really solder well. I resoldered the chip and it fired right up! Thanks! (:
